### ISSUE TITLE
Bug: Discrepancy between query nested filter and sort nested filter

### ISSUE THREAD

**Issue by MacMcIrish** [elastic/elasticsearch#41559]
**Elasticsearch version** (`bin/elasticsearch --version`): 6.3.0

**Plugins installed**: []

**JVM version** (`java -version`): 1.8.0_141

**OS version** (`uname -a` if on a Unix-like system): Linux 4.15.0-47-generic

**Description of the problem including expected versus actual behavior**:

There's a major discrepancy between how query filters behave against how sort filters behave.
I've prepared theoretically equivalent queries below, query1.json and query2.json.

The results of each query are clearly not equivalent, is this a bug in elasticsearch or am I misunderstanding how sort queries should be used?

According to [the docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-sort.html#nested-sorting), the queries should be interchangeable.

> Common case is to repeat the query / filter inside the nested filter or query. By default no nested_filter is active.

Please take note of the `sort` field on each hit in `result2.json`. 

**Steps to reproduce**:

I've named the four entities as according to the order I would expect them to be returned, `first`, `second`, `third`, `fourth`.

Please let me know if there's any more information I can provide.

Script to reproduce:
```bash
#!/usr/bin/env bash

printf 'Deleting\n'
curl elasticsearch:9200/entity -X DELETE
printf '\nCreating\n'
curl elasticsearch:9200/entity -X PUT
printf '\nMapping\n'
curl elasticsearch:9200/entity/_mapping/_doc -X PUT -d @mapping.json -H "Content-Type: application/json"
printf '\nBulk\n'
curl elasticsearch:9200/_bulk -X POST --data-binary @data.json -H "Content-Type: application/x-ndjson"
printf '\n'
curl elasticsearch:9200/entity/_refresh
printf '\n'
curl elasticsearch:9200/entity/_search?pretty -d @query1.json -H "Content-Type: application/json" > result1.json
printf '\n'
curl elasticsearch:9200/entity/_search?pretty -d @query2.json -H "Content-Type: application/json" > result2.json
printf '\n'
```

<details>
<summary>mapping.json</summary>

```json
{
  "properties": {
    "id": {
      "type": "keyword"
    },
    "name": {
      "type": "text"
    },
    "schedule": {
      "properties": {
        "start": {
          "type": "date"
        },
        "end": {
          "type": "date"
        }
      },
      "type": "nested"
    }
  }
}
```

</details>

<details>
<summary>data.json</summary>

```json
{ "index": { "_index": "entity", "_type": "_doc", "_id": "1" } }
{ "name": "first", "schedule": { "start": "2020-01-01", "end": "2020-01-05" } }
{ "index": { "_index": "entity", "_type": "_doc", "_id": "2" } }
{ "name": "second", "schedule": { "start": "2020-01-02", "end": "2020-01-06" } }
{ "index": { "_index": "entity", "_type": "_doc", "_id": "3" } }
{ "name": "third", "schedule": { "start": "2019-01-01", "end": "2019-04-03" } }
{ "index": { "_index": "entity", "_type": "_doc", "_id": "4" } }
{ "name": "fourth", "schedule": { "start": "2019-01-01", "end": "2019-03-03" } }

```
</details>

<details>
<summary>query1.json</summary>

```json
{
  "_source": [
    "id",
    "name",
    "schedule.start",
    "schedule.end"
  ],
  "size": 20,
  "from": 0,
  "sort": [
    {
      "schedule.start": {
        "order": "asc",
        "nested": {
          "path": "schedule",
          "filter": {
            "range": {
              "schedule.start": {
                "gte": "2019-06-06"
              }
            }
          }
        }
      }
    },
    {
      "schedule.end": {
        "order": "desc",
        "nested": {
          "path": "schedule",
          "filter": {
            "range": {
              "schedule.end": {
                "lte": "2019-06-06"
              }
            }
          }
        }
      }
    }
  ]
}
```

</details>
<details>
<summary>query2.json</summary>

```json
{
  "_source": [
    "id",
    "name",
    "schedule.start",
    "schedule.end"
  ],
  "size": 20,
  "from": 0,
  "sort": [
    {
      "schedule.start": {
        "order": "asc",
        "nested": {
          "path": "schedule",
          "filter": {
            "bool": {
              "filter": [
                {
                  "nested": {
                    "path": "schedule",
                    "query": {
                      "bool": {
                        "filter": [
                          {
                            "range": {
                              "schedule.start": {
                                "gte": "2019-06-06"
                              }
                            }
                          }
                        ]
                      }
                    }
                  }
                }
              ]
            }
          }
        }
      }
    },
    {
      "schedule.end": {
        "order": "desc",
        "nested": {
          "path": "schedule",
          "filter": {
            "bool": {
              "filter": [
                {
                  "nested": {
                    "path": "schedule",
                    "query": {
                      "bool": {
                        "filter": [
                          {
                            "range": {
                              "schedule.end": {
                                "lte": "2019-06-06"
                              }
                            }
                          }
                        ]
                      }
                    }
                  }
                }
              ]
            }
          }
        }
      }
    }
  ]
}
```

</details>
<details>
<summary>result1.json</summary>

```json
{
  "took" : 1,
  "timed_out" : false,
  "_shards" : {
    "total" : 5,
    "successful" : 5,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : 4,
    "max_score" : null,
    "hits" : [
      {
        "_index" : "entity",
        "_type" : "_doc",
        "_id" : "1",
        "_score" : null,
        "_source" : {
          "schedule" : {
            "start" : "2020-01-01",
            "end" : "2020-01-05"
          },
          "name" : "first"
        },
        "sort" : [
          1577836800000,
          -9223372036854775808
        ]
      },
      {
        "_index" : "entity",
        "_type" : "_doc",
        "_id" : "2",
        "_score" : null,
        "_source" : {
          "schedule" : {
            "start" : "2020-01-02",
            "end" : "2020-01-06"
          },
          "name" : "second"
        },
        "sort" : [
          1577923200000,
          -9223372036854775808
        ]
      },
      {
        "_index" : "entity",
        "_type" : "_doc",
        "_id" : "3",
        "_score" : null,
        "_source" : {
          "schedule" : {
            "start" : "2019-01-01",
            "end" : "2019-04-03"
          },
          "name" : "third"
        },
        "sort" : [
          9223372036854775807,
          1554249600000
        ]
      },
      {
        "_index" : "entity",
        "_type" : "_doc",
        "_id" : "4",
        "_score" : null,
        "_source" : {
          "schedule" : {
            "start" : "2019-01-01",
            "end" : "2019-03-03"
          },
          "name" : "fourth"
        },
        "sort" : [
          9223372036854775807,
          1551571200000
        ]
      }
    ]
  }
}
```

</details>
<details>
<summary>result2.json</summary>

```json
{
  "took" : 1,
  "timed_out" : false,
  "_shards" : {
    "total" : 5,
    "successful" : 5,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : 4,
    "max_score" : null,
    "hits" : [
      {
        "_index" : "entity",
        "_type" : "_doc",
        "_id" : "4",
        "_score" : null,
        "_source" : {
          "schedule" : {
            "start" : "2019-01-01",
            "end" : "2019-03-03"
          },
          "name" : "fourth"
        },
        "sort" : [
          1546300800000,
          -9223372036854775808
        ]
      },
      {
        "_index" : "entity",
        "_type" : "_doc",
        "_id" : "2",
        "_score" : null,
        "_source" : {
          "schedule" : {
            "start" : "2020-01-02",
            "end" : "2020-01-06"
          },
          "name" : "second"
        },
        "sort" : [
          9223372036854775807,
          -9223372036854775808
        ]
      },
      {
        "_index" : "entity",
        "_type" : "_doc",
        "_id" : "1",
        "_score" : null,
        "_source" : {
          "schedule" : {
            "start" : "2020-01-01",
            "end" : "2020-01-05"
          },
          "name" : "first"
        },
        "sort" : [
          9223372036854775807,
          -9223372036854775808
        ]
      },
      {
        "_index" : "entity",
        "_type" : "_doc",
        "_id" : "3",
        "_score" : null,
        "_source" : {
          "schedule" : {
            "start" : "2019-01-01",
            "end" : "2019-04-03"
          },
          "name" : "third"
        },
        "sort" : [
          9223372036854775807,
          -9223372036854775808
        ]
      }
    ]
  }
}

```

</details>

**Comment by jpountz** [comment]
Actually the 2nd query is invalid: it is written as if the `schedule` nested object had another nested object called `schedule` as well. Closing in favor of #23290.

**Comment by MacMcIrish** [comment]
@jpountz cheers for the link. We'll have to do some further thinking in how we dynamically build our queries.
Just so I understand, that it filters the results properly if I put the same filter in a top level `query` is a bug in itself?

**Comment by simlu** [comment]
@jpountz Thank you very much for your input here. I'm still not entirely sure how to fix the query in a general way. Would you mind proof reading the following query? Is this how you can do double nesting for sort filters? Very much appreciated!

<details> <summary>Spoiler</summary>

```json
{
  "_source": [
    "id",
    "name",
    "schedule.start",
    "schedule.end"
  ],
  "size": 20,
  "from": 0,
  "sort": [
    {
      "schedule.start": {
        "order": "asc",
        "mode": "max",
        "nested": {
          "path": "schedule",
          "filter": {
            "bool": {
              "filter": [
                {
                  "range": {
                    "schedule.start": {
                      "gte": "2019-06-06"
                    }
                  }
                }
              ]
            }
          }
        }
      }
    },
    {
      "schedule.end": {
        "order": "desc",
        "mode": "max",
        "nested": {
          "path": "schedule",
          "filter": {
            "bool": {
              "filter": [
                {
                  "range": {
                    "schedule.end": {
                      "lte": "2019-06-06"
                    }
                  }
                },
                {
                  "nested": {
                    "path": "schedule.subschedule",
                    "query": {
                      "bool": {
                        "filter": [
                          {
                            "range": {
                              "schedule.subschedule.end": {
                                "lte": "2019-06-06"
                              }
                            }
                          }
                        ]
                      }
                    }
                  }
                }
              ]
            }
          }
        }
      }
    },
    {
      "id": {
        "order": "asc"
      }
    }
  ]
}
```

</details>

**Comment by jpountz** [comment]
@MacMcIrish When you start specifying a `nested` element in queries or when sorting, everything below only sees fields that belong to this nested path, other fields become invisible. Your sort object says that it should sort on the `start` property of the nested object `schedule` (so far so good) but then. So far so good, but while you are already in the context of the `schedule` nested object, you use a `nested` query on the `schedule` path in the filter. It makes Elasticsearch look for the nested object `schedule.schedule`, which doesn't exist, so the filter doesn't match any documents. If you look at the response, your sort values are either `9223372036854775807` or `-9223372036854775808`. These values are the maximum and minimum long values, which Elasticsearch uses when values are missing at sorting time so that missing values are sorted last.

@simlu This query looks correct. What I don't know is whether it does what you want. :) So to be clear when sorting on `schedule.end`, it will only consider `schedule` objects whose `end` time is less than or equal to `2019-06-06` AND that have at least one `schedule.subschedule` nested object whose `end` property is less than or equal to `2019-06-06`.

**Comment by simlu** [comment]
@jpountz You seem to imply that double `nested.path` entries need to be provided relative. But from what we've seen they need to be provided absolute:
- [Reference1]( https://discuss.elastic.co/t/querying-a-multiple-level-nested-object/56889/5), scroll to query section in comment
- [Reference2](https://github.com/elastic/elasticsearch/issues/11830#issuecomment-114574181), linked and next comment
- [Reference3](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-sort.html#_nested_sorting_examples), this is slightly different from the sort example above, but shows that nested path should be provided absolute

This is also in line with what we found in our testing.

I would expect this to work the same for sort filtering. So the only issue with the initial query that @MacMcIrish posted was then that the same nesting was specified twice (which was what confused elasticsearch). 

tldr; I might have misunderstood, but it appears that your statement above was incorrect. I'll be working on another test case here that hopefully clarifies this further.

**Comment by simlu** [comment]
Follow-up to my previous comment. To reproduce locally follow instructions @MacMcIrish posted above.

<details> <summary>mapping.json</summary>

```json
{
  "properties": {
    "id": {
      "type": "keyword"
    },
    "name": {
      "type": "text"
    },
    "schedule": {
      "properties": {
        "start": {
          "type": "date"
        },
        "end": {
          "type": "date"
        },
        "sub": {
          "properties": {
            "start": {
              "type": "date"
            },
            "end": {
              "type": "date"
            }
          },
          "type": "nested"
        }
      },
      "type": "nested"
    }
  }
}

```

</details>

<details> <summary>data.json</summary>

```json
{ "index": { "_index": "entity", "_type": "_doc", "_id": "1" } }
{ "name": "first", "schedule": { "start": "2020-01-01", "end": "2020-01-05", "sub": { "start": "2020-01-01", "end": "2020-01-05" } } }
{ "index": { "_index": "entity", "_type": "_doc", "_id": "2" } }
{ "name": "second", "schedule": { "start": "2020-01-02", "end": "2020-01-06", "sub": { "start": "2020-01-02", "end": "2020-01-06" } } }
{ "index": { "_index": "entity", "_type": "_doc", "_id": "3" } }
{ "name": "third", "schedule": { "start": "2019-01-01", "end": "2019-04-03", "sub": { "start": "2019-01-01", "end": "2019-04-03" } } }
{ "index": { "_index": "entity", "_type": "_doc", "_id": "4" } }
{ "name": "fourth", "schedule": { "start": "2019-01-01", "end": "2019-03-03", "sub": { "start": "2019-01-01", "end": "2019-03-03" } } }
```

</details>

<details> <summary>query.json</summary>

```json
{
  "_source": [
    "id",
    "name",
    "schedule.start",
    "schedule.end"
  ],
  "size": 20,
  "from": 0,
  "sort": [
    {
      "schedule.start": {
        "order": "asc",
        "mode": "max",
        "nested": {
          "path": "schedule",
          "filter": {
            "bool": {
              "filter": [
                {
                  "nested": {
                    "path": "schedule.sub",
                    "query": {
                      "bool": {
                        "filter": [
                          {
                            "range": {
                              "schedule.sub.start": {
                                "gte": "2019-06-06"
                              }
                            }
                          }
                        ]
                      }
                    }
                  }
                }
              ]
            }
          }
        }
      }
    },
    {
      "schedule.end": {
        "order": "desc",
        "mode": "max",
        "nested": {
          "path": "schedule",
          "filter": {
            "bool": {
              "filter": [
                {
                  "nested": {
                    "path": "schedule.sub",
                    "query": {
                      "bool": {
                        "filter": [
                          {
                            "range": {
                              "schedule.sub.end": {
                                "lte": "2019-06-06"
                              }
                            }
                          }
                        ]
                      }
                    }
                  }
                }
              ]
            }
          }
        }
      }
    },
    {
      "id": {
        "order": "asc"
      }
    }
  ]
}

```

</details>

<details> <summary>result.json</summary>

```json
{
  "took" : 3,
  "timed_out" : false,
  "_shards" : {
    "total" : 5,
    "successful" : 5,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : 4,
    "max_score" : null,
    "hits" : [
      {
        "_index" : "entity",
        "_type" : "_doc",
        "_id" : "1",
        "_score" : null,
        "_source" : {
          "schedule" : {
            "start" : "2020-01-01",
            "end" : "2020-01-05"
          },
          "name" : "first"
        },
        "sort" : [
          1577836800000,
          -9223372036854775808,
          null
        ]
      },
      {
        "_index" : "entity",
        "_type" : "_doc",
        "_id" : "2",
        "_score" : null,
        "_source" : {
          "schedule" : {
            "start" : "2020-01-02",
            "end" : "2020-01-06"
          },
          "name" : "second"
        },
        "sort" : [
          1577923200000,
          -9223372036854775808,
          null
        ]
      },
      {
        "_index" : "entity",
        "_type" : "_doc",
        "_id" : "3",
        "_score" : null,
        "_source" : {
          "schedule" : {
            "start" : "2019-01-01",
            "end" : "2019-04-03"
          },
          "name" : "third"
        },
        "sort" : [
          9223372036854775807,
          1554249600000,
          null
        ]
      },
      {
        "_index" : "entity",
        "_type" : "_doc",
        "_id" : "4",
        "_score" : null,
        "_source" : {
          "schedule" : {
            "start" : "2019-01-01",
            "end" : "2019-03-03"
          },
          "name" : "fourth"
        },
        "sort" : [
          9223372036854775807,
          1551571200000,
          null
        ]
      }
    ]
  }
}

```

</details>

When I try to make the query relative, as to your suggestion @jpountz, I get `[nested] failed to find nested object under path [sub]`. Would be great if you could clarify! Cheers
